File: matlab-package/brainflow/examples/BandPower.m
BoardShim.set_log_file('brainflow.log');
BoardShim.enable_dev_board_logger();

params = BrainFlowInputParams();
board_shim = BoardShim(int32(BoardIDs.SYNTHETIC_BOARD), params);
board_id = int32(BoardIDs.SYNTHETIC_BOARD);
board_descr = BoardShim.get_board_descr(board_id);
sampling_rate = int32(board_descr.sampling_rate);
board_shim.prepare_session();
board_shim.start_stream(45000, '');
pause(10);
board_shim.stop_stream();
nfft = DataFilter.get_nearest_power_of_two(sampling_rate);
data = board_shim.get_board_data(board_shim.get_board_data_count());
board_shim.release_session();

eeg_channels = board_descr.eeg_channels;
eeg_channel = eeg_channels(3);
original_data = data(eeg_channel, :);
detrended = DataFilter.detrend(original_data, int32(DetrendOperations.LINEAR));
[ampls, freqs] = DataFilter.get_psd_welch(detrended, nfft, nfft / 2, sampling_rate, int32(WindowFunctions.HANNING));
band_power_alpha = DataFilter.get_band_power(ampls, freqs, 7.0, 13.0);
band_power_beta = DataFilter.get_band_power(ampls, freqs, 14.0, 30.0);
ratio = band_power_alpha / band_power_beta;

File: tests/rust/tests/band_power.rs
use approx::assert_relative_eq;
use brainflow::board_shim;
use brainflow::data_filter;
use brainflow::BoardIds;
use brainflow::DetrendOperations;
use brainflow::WindowFunctions;

mod common;

#[test]
fn bandpower() {
    let board_id = BoardIds::SyntheticBoard as i32;
    let eeg_channels = board_shim::get_eeg_channels(board_id).unwrap();
    let sampling_rate = board_shim::get_sampling_rate(board_id).unwrap();
    let nfft = data_filter::get_nearest_power_of_two(sampling_rate).unwrap();

    let mut data = common::board_data(None);

    data_filter::detrend(&mut data[eeg_channels[1]], DetrendOperations::Linear as i32).unwrap();
    let mut psd = data_filter::get_psd_welch(
        &mut data[eeg_channels[1]],
        nfft,
        nfft / 2,
        sampling_rate,
        WindowFunctions::BlackmanHarris as i32,
    )
    .unwrap();
    let band_power_alpha = data_filter::get_band_power(&mut psd, 7.0, 13.0).unwrap();
    let band_power_beta = data_filter::get_band_power(&mut psd, 14.0, 30.0).unwrap();
    assert_relative_eq!(band_power_alpha / band_power_beta, 2500.0, epsilon = 1000.0);
}
